sdjournal/read.go

```diff
@@ -129,63 +148,85 @@ func (r *JournalReader) Close() error {
 	return r.journal.Close()
 }
 
-// Follow synchronously follows the JournalReader, writing each new journal entry to writer. The
-// follow will continue until a single time.Time is received on the until channel.
-func (r *JournalReader) Follow(until <-chan time.Time, writer io.Writer) (err error) {
-
-	// Process journal entries and events. Entries are flushed until the tail or
-	// timeout is reached, and then we wait for new events or the timeout.
-	var msg = make([]byte, 64*1<<(10))
-process:
-	for {
-		c, err := r.Read(msg)
-		if err != nil && err != io.EOF {
-			break process
-		}
+// Follow asynchronously follows the JournalReader it takes in a context to stop the following the Journal, it returns a
+// buffered channel of errors and will stop following the journal on the first given error
+func (r *JournalReader) Follow(ctx context.Context, msgs chan<- map[string]interface{}, filter FollowFilter) <-chan error {
+	errChan := make(chan error, 1)
+	go func() {
+		for {
+			kvMap := make(map[string]interface{})
+			select {
+			case <-ctx.Done():
+				errChan <- ctx.Err()
+				return
+			default:
+				var err error
+				var c int
 
-		select {
-		case <-until:
-			return ErrExpired
-		default:
-			if c > 0 {
-				writer.Write(msg[:c])
-				continue process
-			}
-		}
+				// Advance the journal cursor
+				c, err = r.journal.Next()
 
-		// We're at the tail, so wait for new events or time out.
-		// Holds journal events to process. Tightly bounded for now unless there's a
-		// reason to unblock the journal watch routine more quickly.
-		events := make(chan int, 1)
-		pollDone := make(chan bool, 1)
-		go func() {
-			for {
-				select {
-				case <-pollDone:
+				// An unexpected error
+				if err != nil {
+					errChan <- err
 					return
-				default:
-					events <- r.journal.Wait(time.Duration(1) * time.Second)
+				}
+
+				// We have a new journal entry go over the fields
+				// get the data for what we care about and return
+				r.journal.RestartData()
+				if c > 0 {
+				fields:
+					for {
+						s, err := r.journal.EnumerateData()
+						if err != nil || len(s) == 0 {
+							break fields
+						}
+						s = s[:len(s)]
+						arr := strings.SplitN(s, "=", 2)
+						// if we want the pair,
+						// add it to the map
+						if filter(arr[0], arr[1]) {
+							kvMap[arr[0]] = arr[1]
+						}
+					}
+					msgs <- kvMap
 				}
 			}
-		}()
-
-		select {
-		case <-until:
-			pollDone <- true
-			return ErrExpired
-		case e := <-events:
-			pollDone <- true
-			switch e {
-			case SD_JOURNAL_NOP, SD_JOURNAL_APPEND, SD_JOURNAL_INVALIDATE:
-				// TODO: need to account for any of these?
-			default:
-				log.Printf("Received unknown event: %d\n", e)
+
+			// we're at the tail, so wait for new events or time out.
```

Comment:
I dont think this is true, we need to add a `continue` in the `if c > 0 { }`.
e.g.
```
if c > 0 {
    r.journal.RestartData()
    ...
    continue
}
```

A better way would be put all these into a function to make it cleaner
```
if c > 0 {
    readData()
    continue;
```